vm-allocator/src/system.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2018 The Chromium OS Authors. All rights reserved.
+// Copyright © 2019 Intel Corporation
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use vm_memory::{GuestAddress, GuestUsize};
+
+use crate::address::AddressAllocator;
+use crate::id::IdAllocator;
+
+use libc::{sysconf, _SC_PAGESIZE};
+use std::fmt::{self, Display};
+use std::result;
+use std::sync::{Arc, Mutex};
+
+/// Errors associated with system resources allocation.
+#[derive(Debug)]
+pub enum Error {
+    /// Address is none.
+    NoneAddress,
+    /// Address allocation failed.
+    AddressAllocate(crate::address::Error),
+    /// Id allocation failed.
+    IdAllocate(crate::id::Error),
+}
+
+impl Display for Error {
+    // This trait requires `fmt` with this exact signature.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            NoneAddress => write!(f, "The address being allocated is null"),
+            AddressAllocate(e) => write!(f, "Address being allocated failed, err={}", e),
+            IdAllocate(e) => write!(f, "Address being allocated failed, err={}", e),
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Safe wrapper for `sysconf(_SC_PAGESIZE)`.
+#[inline(always)]
+fn pagesize() -> usize {
+    // Trivially safe
+    unsafe { sysconf(_SC_PAGESIZE) as usize }
+}
+
+/// Manages allocating system resources such as address space and interrupt numbers.
+///
+/// # Example - Use the `SystemAddress` builder.
+///
+/// ```
+/// # use vm_allocator::SystemAllocator;
+/// # use vm_memory::{Address, GuestAddress, GuestUsize};
+///   let mut allocator = SystemAllocator::new(
+///           Some(GuestAddress(0x1000)), Some(0x10000),
+///           GuestAddress(0x10000000), 0x10000000,
+///           5, 24, 1).unwrap();
+///    assert_eq!(allocator.allocate_irq(None).unwrap(), 5);
+///    assert_eq!(allocator.allocate_irq(Some(7)).unwrap(), 7);
+///    assert_eq!(allocator.allocate_mmio_addresses(None, 0x1000).unwrap(), GuestAddress(0x1ffff000));
+///
+/// ```
+#[derive(Clone)]
+pub struct SystemAllocator {
+    io_address_space: Arc<Mutex<Option<AddressAllocator>>>,
```

Comment:
Because port IO address space would be None in some modern VMM, but MMIO address space should exist.
But you're right that the legacy irq resource also needs to be Option. @sameo How do you think?